test/extensions/filters/http/decompressor/decompressor_filter_test.cc

```diff
@@ -0,0 +1,281 @@
+  EXPECT_CALL(*decompressor_factory_, createDecompressor()).Times(0);
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "not-matching"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+  TestUtility::headerMapEqualIgnoreOrder(headers_before_filter, *headers_after_filter);
+
+  expectNoDecompression();
+}
+
+TEST_P(DecompressorFilterTest, NoDecompressionContentEncodingNotCurrent) {
+  EXPECT_CALL(*decompressor_factory_, createDecompressor()).Times(0);
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "gzip,mock"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+  TestUtility::headerMapEqualIgnoreOrder(headers_before_filter, *headers_after_filter);
+
+  expectNoDecompression();
+}
+
+TEST_P(DecompressorFilterTest, NoResponseDecompressionNoTransformPresent) {
+  EXPECT_CALL(*decompressor_factory_, createDecompressor()).Times(0);
+  Http::TestHeaderMapImpl headers_before_filter{
+      {"cache-control", Http::Headers::get().CacheControlValues.NoTransform},
+      {"content-encoding", "mock"},
+      {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+  TestUtility::headerMapEqualIgnoreOrder(headers_before_filter, *headers_after_filter);
+
+  expectNoDecompression();
+}
+
+TEST_P(DecompressorFilterTest, DisableAdvertiseAcceptEncoding) {
+  setUpFilter(R"EOF(
+decompressor_library:
+  typed_config:
+    "@type": "type.googleapis.com/envoy.extensions.compression.gzip.decompressor.v3.Gzip"
+request_direction_config:
+  advertise_accept_encoding: false
+)EOF");
+
+  // Keep the decompressor to set expectations about it
+  auto decompressor = std::make_unique<Compression::Decompressor::MockDecompressor>();
+  auto* decompressor_ptr = decompressor.get();
+  EXPECT_CALL(*decompressor_factory_, createDecompressor())
+      .WillOnce(Return(ByMove(std::move(decompressor))));
+  Http::TestHeaderMapImpl headers_before_filter{{"content-encoding", "mock"},
+                                                {"content-length", "256"}};
+  std::unique_ptr<Http::RequestOrResponseHeaderMap> headers_after_filter =
+      doHeaders(headers_before_filter, false /* end_stream */);
+
+  EXPECT_EQ(nullptr, headers_after_filter->ContentEncoding());
+  EXPECT_EQ(nullptr, headers_after_filter->ContentLength());
+  EXPECT_EQ("chunked", headers_after_filter->TransferEncoding()->value().getStringView());
+  EXPECT_EQ(nullptr, headers_after_filter->get(Http::LowerCaseString{"accept-encoding"}));
+
+  expectDecompression(decompressor_ptr);
+}
+
```

Comment:
I think it might make sense to battle test some of the header parsing logic a bit more as well (i.e., case insensitive headers, spacing, etc.)

Reply:
added more tests